File: packages/shared-js/src/protocol.ts
/**
 * Shared Socket.IO event types used by both the backend (be/) and the
 * frontend network layer (fe/src/net/). Keeping them here ensures the
 * wire format stays in sync across both packages.
 */

/**
 * A player's swap move.
 *
 * Client → server submissions use client-local timestamp values only for
 * diagnostics. Server → client relays set `timestamp` to the server receipt
 * time, so clients can reconcile the chess clock against when the move
 * actually reached the authoritative process.
 */
export interface Move {
  playerId: string;
  r1: number;
  c1: number;
  r2: number;
  c2: number;
  timestamp: number;
}

// ─── v0.9 Flutter-native board-delta protocol ───────────────────────────────

/** Flat row-major board table. Index = row * agreedBoardWidth + col. */
export interface FlatBoardWire {
  boardVersion: number;
  board: number[];
}

/** A generated tile consumed after gravity in deterministic refill order. */
export interface GeneratedTileWire {
  row: number;
  col: number;
  tile: number;
}

/** Hot-path move-resolution packet for the Flutter-native online client. */
export interface BoardDeltaMoveResolvedPayload {
  boardVersion: number;
  playerId: string;
  r1: number;
  c1: number;
  r2: number;
  c2: number;
  serverReceivedAt?: number;
  steps: ResolvedStepWire[];
  generatedTiles: GeneratedTileWire[];
  playerStates?: Record<string, PlayerState>;
  boardHash?: string;
}

/** Start/rejoin snapshot for the Flutter-native online client. */
export interface BoardDeltaMatchFoundPayload extends FlatBoardWire {
  roomId: string;
  opponentId: string;
  myPlayerId: string;
  activePlayerId: string | null;
  mode: "turn_based";
  playerStates: Record<string, PlayerState>;
}

/** Full-board replacement when the settled board has no legal moves. */
export interface BoardReplacedPayload extends FlatBoardWire {
  reason: "no_legal_moves";
  playerStates?: Record<string, PlayerState>;
}

// ─── Server-authoritative PvP types ──────────────────────────────────────────

/**
 * Wire-safe form of a single cascade step from resolveBoardAnimated.
 * All fields are plain numbers/arrays — safe to JSON-serialise over Socket.IO.
 */
export interface ResolvedStepWire {
  /** Cells cleared in this cascade step: [[row, col], ...] */
  matchedCells: [number, number][];
  /** Tiles that fell due to gravity: { col, fromRow, toRow } */
  movements: Array<{ col: number; fromRow: number; toRow: number }>;
  /** Positions that received new tiles: { row, col } */
  newTilePositions: Array<{ row: number; col: number }>;
  /** Grid state after gravity (before refill). */
  afterGravity: number[][];
  /** Grid state after refill — the authoritative snapshot for this cascade. */
  afterRefill: number[][];
  /**
   * Per-player stats AFTER this cascade step's effects (heal/exp/attack/...)
   * have been applied. Lets the client animate HUD bars in lockstep with each
   * cascade flash rather than waiting for the whole chain to settle. Optional
   * during the migration window — older servers / clients don't emit this.
   */
  playerStatesAfter?: Record<string, PlayerState>;
}

/**
 * Internal/server-authoritative resolve payload. During the Flutter-native
 * migration, SocketBridge projects this into a score-free board-delta
 * `move_resolved` hot-path packet for online clients.
 */
export interface MoveResolvedPayload {
  /** The player who made the move. */
  playerId: string;
  r1: number;
  c1: number;
  r2: number;
  c2: number;
  /** Server wall-clock ms when the move event reached the backend. */
  serverReceivedAt?: number;
  /** Monotonic flat-board version after this resolution. */
  boardVersion?: number;
  /** Cascade-by-cascade animation data. Non-empty — 0-match swaps are rejected. */
  steps: ResolvedStepWire[];
  /** Generated replacement tiles in deterministic refill order. */
  generatedTiles?: GeneratedTileWire[];
  /** Final board after all cascades have settled. Authoritative truth. */
  finalGrid: number[][];
  /** RNG state after the resolution — used for snapshot rejoin. */
  rngState: number;
  /** Score delta for this move (added to playerId's running total). */
  pointsEarned: number;
  /** Updated per-player running totals after this move. */
  scores: { [playerId: string]: number };
}

/**
 * Emitted ONLY to the offending socket when a move is rejected.
 */
export interface MoveRejectedPayload {
  reason: "no_match" | "not_your_turn" | "out_of_bounds" | "non_adjacent" | string;
}

/** Payload of the server → client "match_found" event. */
export interface MatchFoundPayload {
  roomId: string;
  /** pve only: legacy seed for local replay. Not sent for turn_based rooms. */
  seed?: number;
  opponentId: string;
  myPlayerId: string;
  firstPlayerId: string;
  activePlayerId?: string | null;
  mode: string;
  /** HMAC-signed token; store in sessionStorage for reconnect after network drop. */
  rejoinToken: string;
  /** turn_based v0.9: flat row-major board snapshot using the agreed board size. */
  boardVersion?: number;
  board?: number[];
  /**
   * pve only: the move log so the client can replay on reconnect. Empty on
   * first connect. Not present for turn_based (server validates privately;
   * clients animate accepted moves locally).
   */
  moves?: Move[];
  /**
   * Initial per-player stats so the HUD can render full HP/Stamina/Mana/Lv
   * bars at match start without waiting for the first turn_changed.
   */
  playerStates?: Record<string, PlayerState>;
}

// ─── Per-player state (replaces the flat `times` field) ─────────────────────

/**
 * Rich per-player state broadcast on every turn change and game-over event.
 *
 * Mirrors `PlayerStats` in `engine/PlayerStats.ts` (the canonical engine
 * shape). All fields travel on the wire so clients can render full HP /
 * stamina / mana / level / exp bars without inferring from deltas.
 *
 * - `stamina` — remaining turn time in ms. Ticks down while the player is
 *   active; match ends when it reaches 0.
 * - `health` — current HP. Match also ends when it reaches 0.
 * - `mana` — current mana (gated abilities, future use).
 * - `lv`, `exp`, `expToNext` — leveling progression. `expToNext = 100 * lv`.
 * - `atk` — current per-attack-tile damage.
 * - `maxHealth` / `maxMana` / `maxStamina` — current caps (maxHealth grows
 *   on level-up; the others are static today but exposed for symmetry).
 */
export interface PlayerState {
  stamina: number;
  maxStamina: number;
  health: number;
  maxHealth: number;
  mana: number;
  maxMana: number;
  lv: number;
  exp: number;
  expToNext: number;
  atk: number;
}

/** Payload of the server → client "turn_changed" event. */
export interface TurnChangedPayload {
  activePlayerId: string;
  /** Server wall-clock ms for the move that caused this turn change. */
  serverReceivedAt?: number;
  /**
   * Per-player state after this turn change. The judge (turn_based) emits the
   * full shape; the legacy pve relay path emits only the stamina-bearing
   * `times` field below — clients must guard for undefined.
   */
  playerStates?: Record<string, PlayerState>;
  /**
   * Legacy stamina-only map used by the pve relay path (`{ [socketId]: ms }`).
   * Kept for backward compatibility; new code should prefer `playerStates`.
   */
  times?: Record<string, number>;
}

/** Why the losing player lost the match. */
export type LoseReason = "time" | "hp";

/** Payload of the server → client "game_over" event. */
export interface GameOverPayload {
  /**
   * Socket ID of the losing player (covers both stamina-out and HP-out
   * cases). Pair with `loserReason` to know which condition triggered.
   * Optional during the migration window — once all emitters set it, drop
   * the `?`.
   */
  loserId?: string;
  /**
   * Why the loser lost. `"time"` ↔ stamina hit zero, `"hp"` ↔ health hit
   * zero (player-stats system).
   */
  loserReason?: LoseReason;
  /** Per-player state at game end. Replaces the old `times` field. */
  playerStates?: Record<string, PlayerState>;
}

/**
 * Payload of the server → client "rejoin_ok" event.
 *
 * For turn_based rooms: contains a one-shot flat board snapshot so the client
 * can restore state instantly without replaying move history.
 *
 * For pve rooms: retains the legacy seed + moves[] shape (unchanged).
 */
export interface RejoinOkPayload {
  roomId: string;
  /**
   * Kept for pve rooms and backward-compat. Not sent for turn_based rooms.
   */
  seed?: number;
  /**
   * pve rooms only: full move history with playerIds remapped to current socket
   * IDs. Not present for turn_based rooms (use board snapshot instead).
   */
  moves?: Move[];
  myPlayerId: string;
  activePlayerId: string | null;
  /** Per-player state at the moment of rejoin. Replaces the old `times` field. */
  playerStates: Record<string, PlayerState>;
  opponentId: string | null;
  /** Fresh token to replace the one stored in sessionStorage. */
  rejoinToken: string;
  /** turn_based v0.9: flat row-major board snapshot using the agreed board size. */
  boardVersion?: number;
  board?: number[];
  /**
   * turn_based only: per-player score totals at the moment of rejoin.
   */
  scores?: { [playerId: string]: number };
}

/** Emitted to the remaining player while their opponent is reconnecting. */
export interface OpponentReconnectingPayload {
  /** How long the server will wait before ending the game (ms). */
  timeoutMs: number;
}

// ─── Auth error codes (T-v0.6-D07) ───────────────────────────────────────────
// Machine-readable codes the client maps to UX messages.
// "AUTH_MISSING_TOKEN" — handshake carried no token at all.
// "AUTH_INVALID_TOKEN" — token present but signature / claims failed verification.
// "AUTH_EXPIRED"       — token signature valid but exp claim has passed.

export type AuthErrorCode =
  | "AUTH_MISSING_TOKEN"
  | "AUTH_INVALID_TOKEN"
  | "AUTH_EXPIRED";

/**
 * Emitted by the server when a socket's token is rejected mid-session.
 * The socket is disconnected immediately after emission.
 * Clients should request a fresh token and reconnect.
 */
export interface AuthTokenRejectedPayload {
  code: AuthErrorCode;
  reason: string;
}

File: apps/frontend/lib/net/protocol.dart
library;

import '../game_core/board.dart';

class PlayerStateDto {
  const PlayerStateDto({
    required this.stamina,
    required this.maxStamina,
    required this.health,
    required this.maxHealth,
    required this.mana,
    required this.maxMana,
    required this.lv,
    required this.exp,
    required this.expToNext,
    required this.atk,
  });

  final int stamina;
  final int maxStamina;
  final int health;
  final int maxHealth;
  final int mana;
  final int maxMana;
  final int lv;
  final int exp;
  final int expToNext;
  final int atk;

  factory PlayerStateDto.fromJson(Map<String, dynamic> json) => PlayerStateDto(
        stamina: _readInt(json, 'stamina'),
        maxStamina: _readInt(json, 'maxStamina'),
        health: _readInt(json, 'health'),
        maxHealth: _readInt(json, 'maxHealth'),
        mana: _readInt(json, 'mana'),
        maxMana: _readInt(json, 'maxMana'),
        lv: _readInt(json, 'lv'),
        exp: _readInt(json, 'exp'),
        expToNext: _readInt(json, 'expToNext'),
        atk: _readInt(json, 'atk'),
      );
}

class FlatBoardDto {
  const FlatBoardDto({
    required this.boardVersion,
    required this.board,
  });

  int get width => defaultBoardWidth;
  int get height => defaultBoardHeight;
  final int boardVersion;
  final List<int> board;

  factory FlatBoardDto.fromJson(Map<String, dynamic> json) {
    final board = _readIntList(json, 'board');
    if (board.length != defaultBoardWidth * defaultBoardHeight) {
      throw FormatException(
        'Flat board length ${board.length} does not match agreed '
        '$defaultBoardWidth x $defaultBoardHeight board',
      );
    }
    return FlatBoardDto(
      boardVersion: _readInt(json, 'boardVersion'),
      board: List.unmodifiable(board),
    );
  }
}

class GeneratedTileDto {
  const GeneratedTileDto({
    required this.row,
    required this.col,
    required this.tile,
  });

  final int row;
  final int col;
  final int tile;

  factory GeneratedTileDto.fromJson(Map<String, dynamic> json) =>
      GeneratedTileDto(
        row: _readInt(json, 'row'),
        col: _readInt(json, 'col'),
        tile: _readInt(json, 'tile'),
      );
}

class BoardDeltaMatchFoundDto extends FlatBoardDto {
  BoardDeltaMatchFoundDto({
    required super.boardVersion,
    required super.board,
    required this.roomId,
    required this.mode,
    required this.activePlayerId,
    required this.myPlayerId,
    required this.opponentId,
    required this.playerStates,
  });

  final String roomId;
  final String mode;
  final String? activePlayerId;
  final String myPlayerId;
  final String opponentId;
  final Map<String, PlayerStateDto> playerStates;

  factory BoardDeltaMatchFoundDto.fromJson(Map<String, dynamic> json) {
    final board = FlatBoardDto.fromJson(json);
    return BoardDeltaMatchFoundDto(
      boardVersion: board.boardVersion,
      board: board.board,
      roomId: json['roomId'] as String,
      mode: json['mode'] as String,
      activePlayerId: json['activePlayerId'] as String?,
      myPlayerId: json['myPlayerId'] as String,
      opponentId: json['opponentId'] as String,
      playerStates: _parsePlayerStates(json['playerStates']),
    );
  }
}

class BoardReplacedDto extends FlatBoardDto {
  BoardReplacedDto({
    required super.boardVersion,
    required super.board,
    required this.reason,
    required this.playerStates,
  });

  final String reason;
  final Map<String, PlayerStateDto> playerStates;

  factory BoardReplacedDto.fromJson(Map<String, dynamic> json) {
    final board = FlatBoardDto.fromJson(json);
    return BoardReplacedDto(
      boardVersion: board.boardVersion,
      board: board.board,
      reason: json['reason'] as String,
      playerStates: _parsePlayerStates(json['playerStates']),
    );
  }
}

class MoveResolvedDto {
  const MoveResolvedDto({
    required this.boardVersion,
    required this.playerId,
    required this.r1,
    required this.c1,
    required this.r2,
    required this.c2,
    required this.steps,
    required this.generatedTiles,
    required this.playerStates,
  });

  final int boardVersion;
  final String playerId;
  final int r1;
  final int c1;
  final int r2;
  final int c2;
  final List<ResolvedStepDto> steps;
  final List<GeneratedTileDto> generatedTiles;
  final Map<String, PlayerStateDto> playerStates;

  factory MoveResolvedDto.fromJson(Map<String, dynamic> json) {
    final steps = (json['steps'] as List<dynamic>)
        .map((raw) => ResolvedStepDto.fromJson(raw as Map<String, dynamic>))
        .toList(growable: false);
    final generatedTiles = (json['generatedTiles'] as List<dynamic>)
        .map((raw) => GeneratedTileDto.fromJson(raw as Map<String, dynamic>))
        .toList(growable: false);

    return MoveResolvedDto(
      boardVersion: _readInt(json, 'boardVersion'),
      playerId: json['playerId'] as String,
      r1: _readInt(json, 'r1'),
      c1: _readInt(json, 'c1'),
      r2: _readInt(json, 'r2'),
      c2: _readInt(json, 'c2'),
      steps: steps,
      generatedTiles: generatedTiles,
      playerStates: _parsePlayerStates(json['playerStates']),
    );
  }
}

class ResolvedStepDto {
  const ResolvedStepDto({
    required this.matchedCells,
    required this.movements,
    required this.newTilePositions,
    required this.afterGravity,
    required this.afterRefill,
    required this.playerStatesAfter,
  });

  final List<BoardCellDto> matchedCells;
  final List<TileMovementDto> movements;
  final List<BoardCellDto> newTilePositions;
  final List<List<int>> afterGravity;
  final List<List<int>> afterRefill;
  final Map<String, PlayerStateDto> playerStatesAfter;

  factory ResolvedStepDto.fromJson(Map<String, dynamic> json) {
    return ResolvedStepDto(
      matchedCells: _parseCellPairs(json['matchedCells']),
      movements: (json['movements'] as List<dynamic>? ?? const [])
          .map((raw) => TileMovementDto.fromJson(raw as Map<String, dynamic>))
          .toList(growable: false),
      newTilePositions: (json['newTilePositions'] as List<dynamic>)
          .map((raw) => BoardCellDto.fromJson(raw as Map<String, dynamic>))
          .toList(growable: false),
      afterGravity: _parseGrid(json['afterGravity']),
      afterRefill: _parseGrid(json['afterRefill']),
      playerStatesAfter: _parsePlayerStates(json['playerStatesAfter']),
    );
  }
}

class TileMovementDto {
  const TileMovementDto({
    required this.col,
    required this.fromRow,
    required this.toRow,
  });

  final int col;
  final int fromRow;
  final int toRow;

  factory TileMovementDto.fromJson(Map<String, dynamic> json) =>
      TileMovementDto(
        col: _readInt(json, 'col'),
        fromRow: _readInt(json, 'fromRow'),
        toRow: _readInt(json, 'toRow'),
      );
}

class BoardCellDto {
  const BoardCellDto({required this.row, required this.col});

  final int row;
  final int col;

  factory BoardCellDto.fromJson(Map<String, dynamic> json) => BoardCellDto(
        row: _readInt(json, 'row'),
        col: _readInt(json, 'col'),
      );
}

class TurnChangedDto {
  const TurnChangedDto({
    required this.activePlayerId,
    required this.playerStates,
  });

  final String activePlayerId;
  final Map<String, PlayerStateDto> playerStates;

  factory TurnChangedDto.fromJson(Map<String, dynamic> json) => TurnChangedDto(
        activePlayerId: json['activePlayerId'] as String,
        playerStates: _parsePlayerStates(json['playerStates']),
      );
}

class MoveRejectedDto {
  const MoveRejectedDto({required this.reason});

  final String reason;

  factory MoveRejectedDto.fromJson(Map<String, dynamic> json) =>
      MoveRejectedDto(reason: json['reason']?.toString() ?? 'rejected');
}

class GameOverDto {
  const GameOverDto({
    required this.loserId,
    required this.loserReason,
    required this.playerStates,
  });

  final String? loserId;
  final String? loserReason;
  final Map<String, PlayerStateDto> playerStates;

  factory GameOverDto.fromJson(Map<String, dynamic> json) => GameOverDto(
        loserId: json['loserId'] as String?,
        loserReason: json['loserReason'] as String?,
        playerStates: _parsePlayerStates(json['playerStates']),
      );
}

Map<String, PlayerStateDto> _parsePlayerStates(Object? raw) {
  if (raw == null) return const {};
  final map = raw as Map<String, dynamic>;
  return Map.unmodifiable(map.map(
    (key, value) => MapEntry(
      key,
      PlayerStateDto.fromJson(value as Map<String, dynamic>),
    ),
  ));
}

List<BoardCellDto> _parseCellPairs(Object? raw) {
  if (raw == null) return const [];
  return List<BoardCellDto>.unmodifiable((raw as List<dynamic>).map((cell) {
    final pair = cell as List<dynamic>;
    if (pair.length != 2) {
      throw const FormatException('matchedCells entry must have row and col');
    }
    return BoardCellDto(
      row: _readIntValue(pair[0]),
      col: _readIntValue(pair[1]),
    );
  }));
}

List<List<int>> _parseGrid(Object? raw) {
  final rows = raw as List<dynamic>;
  return List<List<int>>.unmodifiable(
    rows.map((row) => List<int>.unmodifiable(
          (row as List<dynamic>).map(_readIntValue),
        )),
  );
}

int _readInt(Map<String, dynamic> json, String key) {
  if (!json.containsKey(key) || json[key] == null) {
    throw FormatException('Missing required integer field "$key"');
  }
  return _readIntValue(json[key]);
}

int _readIntValue(Object? value) {
  if (value is int) return value;
  if (value is num && value % 1 == 0) return value.toInt();
  throw FormatException('Expected integer, got ${value.runtimeType}');
}

List<int> _readIntList(Map<String, dynamic> json, String key) {
  final raw = json[key];
  if (raw is! List<dynamic>) {
    throw FormatException('Missing required integer array "$key"');
  }
  return raw.map(_readIntValue).toList(growable: false);
}
